mlflow/pyspark/ml/__init__.py

```diff
@@ -0,0 +1,193 @@
+            self._parent = _SparkTrainingSession._session_stack[-1]
+            self.allow_children = (
+                _SparkTrainingSession._session_stack[-1].allow_children and self.allow_children
+            )
+        _SparkTrainingSession._session_stack.append(self)
+        return self
+
+    def __exit__(self, tp, val, traceback):
+        _SparkTrainingSession._session_stack.pop()
+
+    def should_log(self):
+        """
+        Returns True when at least one of the following conditions satisfies:
+
+        1. This session is the root session.
+        2. The parent session allows autologging and its class differs from this session's class.
+        """
+        return (self._parent is None) or (
+            self._parent.allow_children and self._parent.clazz != self.clazz
+        )
+
+
+@experimental
+@autologging_integration(mlflow_spark.FLAVOR_NAME)
+def autolog(
+    log_models=True,
+    disable=False,
+    exclusive=False,
+    disable_for_unsupported_versions=False,
+    silent=False,
+):  # pylint: disable=unused-argument
+    """
+    Enables (or disables) and configures autologging for pyspark ml estimators.
+
+    :param log_models:
+    :param disable:
+    :param exclusive:
+    :param disable_for_unsupported_versions:
+    :param silent:
+    :return:
+    """
+    from mlflow.utils.validation import (
+        MAX_PARAMS_TAGS_PER_BATCH,
+        MAX_PARAM_VAL_LENGTH,
+        MAX_ENTITY_KEY_LENGTH,
+    )
+    from pyspark.ml.base import Estimator
+
+    def _log_pretraining_metadata(estimator, params):
+
+        if params and isinstance(params, dict):
+            estimator = estimator.copy(params)
+
+        # Chunk model parameters to avoid hitting the log_batch API limit
+        for chunk in chunk_dict(
+                _get_estimator_param_map(estimator),
+                chunk_size=MAX_PARAMS_TAGS_PER_BATCH,
+        ):
+            truncated = truncate_dict(chunk, MAX_ENTITY_KEY_LENGTH, MAX_PARAM_VAL_LENGTH)
+            try_mlflow_log(mlflow.log_params, truncated)
+
+        try_mlflow_log(mlflow.set_tags, _get_estimator_info_tags(estimator))
+
+    def _log_posttraining_metadata(estimator, spark_model, params):
+        if log_models and _should_log_model(estimator, spark_model):
+            # TODO:
+            #  1. check whitelist when log models
+            #  2. support model signature ?
+            try_mlflow_log(
+                mlflow_spark.log_model,
+                spark_model,
+                artifact_path="model",
+            )
+
+    def fit_mlflow(original, self, params=None):
+        _log_pretraining_metadata(self, params)
+        spark_model = original(self, params)
+        _log_posttraining_metadata(self, spark_model, params)
+        return spark_model
+
+    def patched_fit(original, self, *args, **kwargs):
+        params = None
+        if len(args) > 0:
+            params = args[0]
+        elif 'params' in kwargs:
+            params = kwargs['params']
```

Comment:
We still need the ugly code because the patched fit have to be defined like `def patched_fit(original, self, *args, **kwargs):` otherwise safety check in mlflow patching will raise error.